napi-inl.h

```diff
@@ -865,47 +942,65 @@ inline const T* TypedArray_<T,A>::Data() const {
 // Function class
 ////////////////////////////////////////////////////////////////////////////////
 
-inline Function Function::New(napi_env env,
-                              VoidFunctionCallback cb,
-                              const char* utf8name,
-                              void* data) {
-  // TODO: Delete when the function is destroyed
-  CallbackData* callbackData = new CallbackData(cb, data);
+// Helpers to handle functions exposed from C++.
+namespace helper {
 
-  // TODO: set the function name
-  napi_value value;
-  napi_status status = napi_create_function(
-    env, utf8name, VoidFunctionCallbackWrapper, callbackData, &value);
-  if (status != napi_ok) throw Error::New(env);
-  return Function(env, value);
-}
+template <typename Callable, typename Return>
+struct CallbackData {
+  static inline
+  napi_value Wrapper(napi_env env, napi_callback_info info) {
+    try {
+      CallbackInfo callbackInfo(env, info);
+      CallbackData* callbackData =
+        static_cast<CallbackData*>(callbackInfo.Data());
+      return callbackData->callback(callbackInfo);
+    }
+    NAPI_RETHROW_JS_ERROR(env)
+  }
+
+  Callable callback;
+};
+
+template <typename Callable>
+struct CallbackData<Callable, void> {
+  static inline
+  napi_value Wrapper(napi_env env, napi_callback_info info) {
+    try {
+      CallbackInfo callbackInfo(env, info);
+      CallbackData* callbackData =
+        static_cast<CallbackData*>(callbackInfo.Data());
+      callbackData->callback(callbackInfo);
+    }
+    NAPI_RETHROW_JS_ERROR(env)
+  }
 
+  Callable callback;
+};
+
+}  // namespace helper
+
+template<typename Callable>
 inline Function Function::New(napi_env env,
-                              FunctionCallback cb,
-                              const char* utf8name,
-                              void* data) {
```

Comment:
`real_data` is the data argument supplied to function::New, stored on an `Object` (as hidden value or such for V8), which is passed as the `data` argument to `v8::Function::New` or equivalent. Then this gets unwrapped before calling the callback the user sees.
However, if this was not working previously, I guess it is better to leave this as-is and address `data` in a separate PR. Maybe there was a plan for it.

For future reference, perhaps the actual code I was thinking of makes this comprehensible
```c++
Factory<v8::Function>::return_t
Factory<v8::Function>::New( FunctionCallback callback
                          , v8::Local<v8::Value> data) {
  v8::Isolate *isolate = v8::Isolate::GetCurrent();
  v8::EscapableHandleScope scope(isolate);
  v8::Local<v8::ObjectTemplate> tpl = v8::ObjectTemplate::New(isolate);
  tpl->SetInternalFieldCount(imp::kFunctionFieldCount);
  v8::Local<v8::Object> obj = NewInstance(tpl).ToLocalChecked();

  obj->SetInternalField(
      imp::kFunctionIndex
    , v8::External::New(isolate, reinterpret_cast<void *>(callback)));

  v8::Local<v8::Value> val = v8::Local<v8::Value>::New(isolate, data);

  if (!val.IsEmpty()) {
    obj->SetInternalField(imp::kDataIndex, val);
  }

  return scope.Escape(v8::Function::New( isolate
                          , imp::FunctionCallbackWrapper
                          , obj));
}
...
static
void FunctionCallbackWrapper(const v8::FunctionCallbackInfo<v8::Value> &info) {
  v8::Local<v8::Object> obj = info.Data().As<v8::Object>();
  FunctionCallback callback = reinterpret_cast<FunctionCallback>(
      reinterpret_cast<intptr_t>(
          obj->GetInternalField(kFunctionIndex).As<v8::External>()->Value()));
  FunctionCallbackInfo<v8::Value>
      cbinfo(info, obj->GetInternalField(kDataIndex));
  callback(cbinfo);
}
```